Add remediation tool to fix incorrect analytics data from canceled authorizations

mgascam opened the pull request:
## Summary

Fixes #WOOPMNT-5550

Adds a remediation tool to fix incorrect analytics data caused by a bug that existed between April 2023 and November 2025. When payment authorizations were canceled (not captured), the system incorrectly:

1. Recorded transaction fees (`_wcpay_transaction_fee` and `_wcpay_net` metadata)
2. Created WCPay refund objects for the full order amount  
3. Set order status to "refunded" instead of "cancelled"

This caused **negative values in WooCommerce Analytics** for affected stores using manual capture.

## Changes

| File | Purpose |
|------|---------|
| `includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php` | Core remediation logic with batch processing via Action Scheduler |
| `includes/notes/class-wc-payments-notes-canceled-auth-remediation.php` | WC Admin Inbox Note to notify affected merchants |
| `includes/class-wc-payments.php` | Registers remediation class and inbox note |
| `includes/class-wc-payments-status.php` | Adds tool to WooCommerce > Status > Tools |
| `tests/unit/migrations/test-class-wc-payments-remediate-canceled-auth-fees.php` | 48 unit tests |
| `tests/unit/notes/test-class-wc-payments-notes-canceled-auth-remediation.php` | 2 unit tests |

## How It Works

1. **Detection**: Finds orders with `_intention_status = canceled` created after April 2023 that have incorrect fee metadata, WCPay refunds, or "refunded" status
2. **Remediation**: Removes fee metadata, deletes WCPay refunds (preserves manual refunds), changes status to "cancelled"
3. **Analytics Sync**: Cleans up `wp_wc_order_stats` table and triggers WooCommerce Analytics sync
4. **Audit Trail**: Adds detailed order note documenting all changes

## Safety Measures

- ✅ Only affects orders with `_intention_status = canceled`
- ✅ Only deletes refunds with `_wcpay_refund_id` metadata (preserves manual refunds)
- ✅ Supports both HPOS and legacy CPT storage
- ✅ Processes in adaptive batches (10-100 orders) to avoid timeouts
- ✅ Safe to run multiple times
- ✅ Detailed logging via `wc_get_logger()`

## Testing Instructions

### Setup - Create test data

1. Use the [test data generator script](https://gist.github.com/mgascam/79f0441b069ddff70ad2174e7831d7c2)
2. Place it in your WordPress root and run: `wp eval-file test-orders.php`
3. This creates 6 test orders covering all affected scenarios

### Test orders created

| # | Total | Status | Scenario | Should Remediate? |
|---|-------|--------|----------|-------------------|
| 1 | $50 | cancelled | Fee metadata only | ✅ Yes |
| 2 | $75 | cancelled | WCPay refund only ($75) | ✅ Yes |
| 3 | $120 | cancelled | Fee metadata + WCPay refund ($120) | ✅ Yes |
| 4 | $45 | refunded | Incorrect "refunded" status | ✅ Yes |
| 5 | $30 | cancelled | Before bug date (Feb 2023) | ❌ No |
| 6 | $80 | completed | Succeeded intent (legitimate order) | ❌ No |

### Verify the problem exists (before remediation)

1. Go to **WooCommerce > Analytics > Revenue** (filter to today)
2. You should see:

| Metric | Expected Value | Explanation |
|--------|----------------|-------------|
| Gross Sales | $80 | Only order 6 (completed) counts |
| Returns | $195 | Phantom refunds: $75 (order 2) + $120 (order 3) |
| Net Sales | **-$115** | $80 - $195 = negative\! |

### Run the remediation

1. Go to **WooCommerce > Status > Tools**
2. Find "Fix canceled authorization analytics" tool
3. Click "Run" and confirm the prompt
4. Refresh page to see progress (Running → Completed)

### Verify the fix (after remediation)

1. Tool shows "Completed" with processed/remediated counts
2. Go to **WooCommerce > Analytics > Revenue** (filter to today)
3. You should see:

| Metric | Expected Value | Explanation |
|--------|----------------|-------------|
| Gross Sales | $80 | Order 6 (unchanged) |
| Returns | $0 | Phantom refunds deleted |
| Net Sales | **$80** | Correct\! |

4. Check affected orders (1-4):
   - Fee metadata (`_wcpay_transaction_fee`, `_wcpay_net`) removed
   - WCPay refunds deleted
   - Order 4 status changed from "refunded" to "cancelled"
   - Order note documents all changes

5. Check unaffected orders (5-6):
   - Order 5: Still has fee metadata (before bug date, not remediated)
   - Order 6: Still has fee metadata (succeeded intent, legitimate)

### Cleanup

```bash
wp option delete wcpay_test_data_created
wp option delete wcpay_fee_remediation_status
wp option delete wcpay_fee_remediation_stats
```

---

- [x] Run `npm run changelog` to add a changelog file
- [x] Covered with tests (50 unit tests)
- [x] Tested on mobile (N/A - admin tool only)

**Post merge**

- [ ] _QA Testing Not Applicable_ (analytics fix, no user-facing payment flow changes)

mgascam committed:
Add remediation for canceled authorization fees

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
feat: add HPOS support for affected orders remediation

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
Address copilot comments

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,508 @@
+/**
+ * Remediates incorrect transaction fees and refunds for canceled authorizations.
+ *
+ * Between April 2023 (commit 2d13751) and November 2025 (commit 3058f59ad),
```
Date is correct

mgascam commented on includes/class-wc-payments.php:
```diff
@@ -712,6 +723,7 @@ function () {
 		require_once __DIR__ . '/migrations/class-erase-bnpl-announcement-meta.php';
 		require_once __DIR__ . '/migrations/class-erase-deprecated-flags-and-options.php';
 		require_once __DIR__ . '/migrations/class-manual-capture-payment-method-settings-update.php';
+		require_once __DIR__ . '/migrations/class-wc-payments-remediate-canceled-auth-fees.php';
```
Done

mgascam commented on includes/class-wc-payments.php:
```diff
@@ -1528,6 +1540,34 @@ public static function update_plugin_version() {
+		$plugin_data = get_plugin_data( WCPAY_PLUGIN_FILE );
+		$new_version = $plugin_data['Version'];
+
+		// Trigger version gate check.
```
Done

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,508 @@
+			AND p.post_date >= %s
+			AND pm_status.meta_key = '_intention_status'
+			AND pm_status.meta_value = %s
+			AND (pm_fee.meta_key = '_wcpay_transaction_fee' OR pm_fee.meta_key = '_wcpay_net')
```
Both can be incorrectly set

mgascam commented on includes/class-wc-payments.php:
```diff
@@ -1528,6 +1540,34 @@ public static function update_plugin_version() {
+			return;
+		}
+
+		// Get new version.
```
Fixed

mgascam replied:
Fixed

mgascam committed:
feat: enhance remediation logic for canceled authorization fees and improve unit tests

mgascam committed:
feat: implement admin notice and remediation tool for canceled authorization fees

mgascam committed:
refactor: remove unused plugin update method and clean up test imports

mgascam committed:
feat: enhance remediation tool for canceled authorization fees with dynamic button text and status description

mgascam committed:
feat: update remediation logic to change order status from 'refunded' to 'cancelled' and add corresponding unit tests

mgascam committed:
feat: add refund stats cleanup logic and corresponding unit tests for remediation process

mgascam committed:
feat: add confirmation message for canceled authorization remediation process

mgascam committed:
feat: remove deprecated admin notice for canceled authorization fee remediation and add new inbox note

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
feat: remove action for plugin update on upgrader process completion

mgascam commented on includes/class-wc-payments.php:
```diff
@@ -352,6 +359,7 @@ public static function init() {
 		add_action( 'admin_init', [ __CLASS__, 'add_woo_admin_notes' ] );
 		add_action( 'admin_init', [ __CLASS__, 'remove_deprecated_notes' ] );
 		add_action( 'init', [ __CLASS__, 'install_actions' ] );
+		add_action( 'upgrader_process_complete', [ __CLASS__, 'on_plugin_update' ], 10, 2 );
```
Done [f93e62e](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/f93e62ee13d0d402e5fe6c8de5efdff6e96baf9f)

mgascam committed:
fix: update SQL queries to use IN clause for transaction fee meta keys and correct post type placeholder

mgascam commented on includes/admin/class-wc-payments-admin-notice-canceled-auth-remediation.php:
```diff
@@ -0,0 +1,156 @@
+					} );
+				} );
+			} );
+		</script>
```
No longer relevant as we are now using an Inbox notification instead

mgascam committed:
feat: enhance refund deletion message to include deleted refund IDs

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+			$order->delete_meta_data( '_wcpay_transaction_fee' );
+			$order->delete_meta_data( '_wcpay_net' );
+			$order->delete_meta_data( '_wcpay_refund_id' );
+			$order->delete_meta_data( '_wcpay_refund_status' );
```
Done

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+			AND o.date_created_gmt >= %s
+			AND pm_status.meta_key = '_intention_status'
+			AND pm_status.meta_value = %s
+			AND (pm_fee.meta_key = '_wcpay_transaction_fee' OR pm_fee.meta_key = '_wcpay_net')
```
Fixed 3f7e799

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+			FROM {$wpdb->posts} p
+			INNER JOIN {$wpdb->postmeta} pm_status ON p.ID = pm_status.post_id
+			INNER JOIN {$wpdb->postmeta} pm_fee ON p.ID = pm_fee.post_id
+			WHERE p.post_type IN ('shop_order', 'shop_order_placehold')
```
Done

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+	 * @return void
+	 */
+	private function schedule_next_batch(): void {
+		if ( ! function_exists( 'as_schedule_single_action' ) ) {
```
Done

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+	 *
+	 * @return void
+	 */
+	public function process_batch(): void {
+		// Check if already complete.
+		if ( $this->is_complete() ) {
+			return;
+		}
+
+		$start_time = microtime( true );
+		$batch_size = $this->get_batch_size();
+		$orders     = $this->get_affected_orders( $batch_size );
+
+		// If no orders found, mark as complete.
+		if ( empty( $orders ) ) {
+			$this->mark_complete();
+			$this->log_completion();
+			$this->cleanup();
+			return;
+		}
+
+		// Process each order.
+		foreach ( $orders as $order ) {
+			$this->increment_stat( 'processed' );
+
+			if ( $this->remediate_order( $order ) ) {
+				$this->increment_stat( 'remediated' );
+				wc_get_logger()->info(
+					sprintf( 'Remediated order %d', $order->get_id() ),
+					[ 'source' => 'wcpay-fee-remediation' ]
+				);
+			} else {
+				$this->increment_stat( 'errors' );
+			}
+
+			// Update last order ID.
+			$this->update_last_order_id( $order->get_id() );
+		}
```
Action Scheduler handles this

mgascam commented on includes/class-wc-payments.php:
```diff
@@ -1509,6 +1520,43 @@ public static function update_plugin_version() {
+		}
+
+		// Trigger version gate check.
+		self::$fee_remediation->maybe_schedule_remediation( $new_version );
```
The remediation will be triggered manually, not upon upgrade.

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+		// Skip if previous version was before bug introduction.
+		if ( version_compare( $previous_version, self::BUG_INTRODUCED_VERSION, '<' ) ) {
+			return;
+		}
```
No longer relevant, the remediation is no longer run upon upgrade

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,585 @@
+			// Delete all refund objects.
+			foreach ( $refunds as $refund ) {
+				$refund->delete( true ); // Force delete, bypass trash.
+			}
```
get_wcpay_refunds() method  handles this

mgascam committed:
Add changelog

mgascam commented:
Hi @elazzabi I requested your review because you worked on the fix and have some context already.

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
fix: enhance confirmation message for canceled authorization remediation and clarify analytics sync process

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,682 @@
+			$order->save();
+
+			// Trigger analytics sync to update wc_order_stats table. This is necessary because
+			// WooCommerce doesn't automatically sync when refunds are deleted (see issue #1073).
```
Fixed [2fbe206](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/2fbe206010140544234bd92d9bf570b7805ff587)

mgascam commented on includes/class-wc-payments-status.php:
```diff
@@ -94,6 +94,14 @@ public function debug_tools( $tools ) {
+					'name'     => __( 'Fix canceled authorization analytics', 'woocommerce-payments' ),
+					'button'   => $this->get_remediation_button_text(),
+					'desc'     => $this->get_remediation_description(),
+					'confirm'  => __( 'This will update order metadata and delete incorrect refund records for affected orders. Make sure you have a recent backup before proceeding. Continue?', 'woocommerce-payments' ),
```
Fixed [2fbe206](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/2fbe206010140544234bd92d9bf570b7805ff587)

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,682 @@
+		$sql     .= ' ORDER BY o.id ASC LIMIT %d';
+		$params[] = $limit;
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+		$order_ids = $wpdb->get_col( $wpdb->prepare( $sql, $params ) );
+
```
SQL is properly prepared with $wpdb->prepare()

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,682 @@
+		$placeholders = implode( ', ', array_fill( 0, count( $refund_ids ), '%d' ) );
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.InterpolatedNotPrepared, WordPress.DB.PreparedSQLPlaceholders.UnfinishedPrepare
+		$wpdb->query( $wpdb->prepare( "DELETE FROM {$wpdb->prefix}wc_order_stats WHERE order_id IN ({$placeholders})", $refund_ids ) );
```
SQL is properly prepared with $wpdb->prepare()

mgascam commented on includes/notes/class-wc-payments-notes-canceled-auth-remediation.php:
```diff
@@ -0,0 +1,112 @@
+	 * @return bool
+	 */
+	private static function has_affected_orders() {
+		include_once WCPAY_ABSPATH . 'includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php';
+		$remediation = new WC_Payments_Remediate_Canceled_Auth_Fees();
+		return $remediation->has_affected_orders();
```
Only runs on admin init for inbox notes, not frequently.

mgascam commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,682 @@
+		$sql     .= ' ORDER BY p.ID ASC LIMIT %d';
+		$params[] = $limit;
+
+		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
+		$order_ids = $wpdb->get_col( $wpdb->prepare( $sql, $params ) );
```
SQL is properly prepared with $wpdb->prepare()

mgascam committed:
Fix linter error

mgascam committed:
fix: update remediation process to trigger WooCommerce refund deletion hook and remove unused method

mgascam committed:
feat: add dry run functionality for canceled authorization fee remediation

frosso commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,847 @@
+	/**
+	 * Version where bug was introduced (5.8.0 = April 2023, commit 2d13751).
+	 */
+	const BUG_INTRODUCED_VERSION = '5.8.0';
```
Is this needed as part of a const? Or can it just be a comment on the class?

mgascam replied:
I should've removed it, done here [2d3406b](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/2d3406b014a7d562d4fce2489d5b546a9c83c032) thanks!

frosso commented on includes/class-wc-payments-status.php:
```diff
@@ -155,6 +170,214 @@ public function delete_test_orders() {
+		}
+
+		include_once WCPAY_ABSPATH . 'includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php';
+		return as_has_scheduled_action( WC_Payments_Remediate_Canceled_Auth_Fees::ACTION_HOOK );
```
I noticed that here, we only checked if `WC_Payments_Remediate_Canceled_Auth_Fees::ACTION_HOOK` is scheduled. But on the `schedule_canceled_auth_dry_run` method, we check both `WC_Payments_Remediate_Canceled_Auth_Fees::ACTION_HOOK` and `WC_Payments_Remediate_Canceled_Auth_Fees::DRY_RUN_ACTION_HOOK`.

Could this lead to some inconsistent UI state? I'm unsure if we need to check for the "dry run" action here, as well 🤷

mgascam replied:
Good catch ⚾  I'm also checking `WC_Payments_Remediate_Canceled_Auth_Fees::DRY_RUN_ACTION_HOOK ` here 
[929394c](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/929394c1caf5877f1107bf1780ba802eabffb1a3) thanks!

frosso reviewed (commented):
Excellent work with the addition of the dry run option!

I ran all the steps for the "Fix canceled authorization analytics", and they worked flawlessly!

<img width="308" height="690" alt="Image" src="https://github.com/user-attachments/assets/542fb0e8-0ff7-4918-b674-1d697772cfc7" />
<img width="2337" height="976" alt="Image" src="https://github.com/user-attachments/assets/8a7aef85-1435-49d8-8bdc-ff0356947d68" />
<img width="1756" height="91" alt="Image" src="https://github.com/user-attachments/assets/3f3f28e1-0dbf-44a9-98f7-447397b2a50c" />
<img width="2358" height="1006" alt="Image" src="https://github.com/user-attachments/assets/52ad0d89-ae8d-4820-949c-16ddf0abd92f" />


I later reset the DB with your instructions, and launched the "Dry run" option (I know it's uncommon to _first_ launch the full action and _then_ the dry run).
I stumbled into a state where I couldn't run the full action, anymore. But the `wcpay-fee-remediation` log was created:

<img width="1098" height="428" alt="Image" src="https://github.com/user-attachments/assets/40f2f772-cff0-4e8d-8b3b-09c099e6949f" />
<img width="2364" height="238" alt="Image" src="https://github.com/user-attachments/assets/ef3778bf-220e-429b-ba23-17cc03493497" />

Is this due to my testing approach? 👀

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
feat: add cleanup method for dry run completion in fee remediation

mgascam committed:
fix: clarify bug introduction version in documentation

mgascam committed:
fix: enhance remediation status check to include dry run actions

mgascam commented:
> I later reset the DB with your instructions, and launched the "Dry run" option (I know it's uncommon to _first_ launch the full action and _then_ the dry run).
> I stumbled into a state where I couldn't run the full action, anymore. But the `wcpay-fee-remediation` log was created:

Thanks for the review @frosso. This was indeed a bug. The issue: The dry run was calling `mark_complete() ` when finished, which set `wcpay_fee_remediation_status = 'completed'`. This blocked the full action from running afterward since both methods `check is_complete()` before proceeding.

I pushed a fix [3fa05b8](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/3fa05b8172b87b6e66b723757bbbc86510d37353). Now:

- Dry run → completes → status reset → Full action can run ✓
- Full action in progress → Dry run blocked ✓
- Dry run in progress → Full action blocked ✓

I think I addressed all your comments so it's ready for a new review. I'm also pinging @elazzabi since he kindly offered to help with the reviews 🙏

frosso commented on includes/migrations/class-wc-payments-remediate-canceled-auth-fees.php:
```diff
@@ -0,0 +1,857 @@
+		$default = [
+			'processed'  => 0,
+			'remediated' => 0,
+			'skipped'    => 0,
```
Is this `skipped` stat changed anywhere during the process?

mgascam replied:
The plan was to keep track of skipped orders, but never implemented it 😅 . Now I think it 's a bit of an overkill.  I think we can trust the SQL query to only return orders that need work. 
Removed it here [16c0140](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/16c0140478559448e51f4c26e4a48152d2708a40) thanks!

frosso reviewed (approved):
Thank you! I did a manual run following the instructions:

<img width="1372" height="959" alt="Image" src="https://github.com/user-attachments/assets/c78fbdf5-1f2b-4918-8aa7-52fbffcb9f47" />

After:

<img width="2336" height="463" alt="Image" src="https://github.com/user-attachments/assets/9399ee70-8ad9-4ec3-b8d4-bbb7f2678e5f" />
<img width="1149" height="436" alt="Image" src="https://github.com/user-attachments/assets/deedd88a-ea28-479b-8ee0-5ede6ea3509a" />
<img width="302" height="294" alt="Image" src="https://github.com/user-attachments/assets/e35ffe0e-ba8b-42c6-aac9-c36082457d4f" />
<img width="295" height="714" alt="Image" src="https://github.com/user-attachments/assets/762c5624-e296-4717-9fb3-f5861bb6e229" />
<img width="2360" height="767" alt="Image" src="https://github.com/user-attachments/assets/2058f185-5b0e-496a-a72f-a8bc2e1cccd6" />

As you can see, the logs include some markup, but it's not a huge deal 🤷

mgascam committed:
fix: remove unused skipped stats

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
fix: sanitize output in dry run remediation log

elazzabi reviewed (changes requested):
I tested the script and steps locally and in multiple JN sites, I always get the same behavior.

After executing the script, I confirm that the analytics show a negative number

<img width="2904" height="1180" alt="Image" src="https://github.com/user-attachments/assets/09bc9b69-6bf4-483a-8c2e-ac732d327d0a" />

After running the remediation tool, I can still see a negative number (just smaller)

<img width="2908" height="1132" alt="Image" src="https://github.com/user-attachments/assets/09e54944-7713-4fbe-8802-6c05861924aa" />

And the logs show that everything is executed

<img width="1160" height="596" alt="Image" src="https://github.com/user-attachments/assets/3c042ec3-907d-4522-b4b2-0c0072d317ff" />

mgascam committed:
fix: ensure order stats are deleted before refund deletion in remediation process

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam commented:
> I tested the script and steps locally and in multiple JN sites, I always get the same behavior.

Thanks for testing @elazzabi! There were a couple of issues I fixed:

- There was a bug. Calling `$refund->delete` was not enough to clean the analytics data. I changed the approach in [e345bb9](https://github.com/Automattic/woocommerce-payments/pull/11140/commits/e345bb9f0d39f4e2c736a1369609dfbea8e4ddcd)
- The test script was creating a test order for the scenario of a partial refund of an authorization. This is not possible to happen and it was only causing confusion, so I edited the script to remove it.

I tested locally from scratch and this is what I see in the analytics:

**Before running the remediation tool**

<img width="1841" height="867" alt="Screenshot 2025-12-24 at 13 07 28" src="https://github.com/user-attachments/assets/035afe4f-8190-4d3f-b271-30f083362a6e" />


<img width="1842" height="757" alt="Screenshot 2025-12-24 at 13 07 37" src="https://github.com/user-attachments/assets/e44ed64c-d74c-41f2-9bc9-c1a9f83a99a9" />

**After running the remediation tool**

<img width="1828" height="754" alt="Screenshot 2025-12-24 at 13 09 44" src="https://github.com/user-attachments/assets/7c8363c7-945e-421e-bb01-bec83f47a070" />

<img width="1833" height="794" alt="Screenshot 2025-12-24 at 13 09 57" src="https://github.com/user-attachments/assets/5c599eae-88aa-4151-b36e-c47758f1fd1f" />

The refunds are no longer appearing in the analytics after the remediation tool run. I updated the testing instructions for better clarity of the expected values. **Before**: Gross Sales $80, Returns $195, Net Sales -$115 **After**: Gross Sales $80, Returns $0, Net Sales $80


It is ready for a new review round when you get a moment, please remember to use the latest version of the test script! Apologies for the back and forth and thanks again! 🙏

mgascam committed:
fix: ensure next batch is always scheduled for fee remediation process

elazzabi reviewed (approved):
I tested with the new changes.

Before executing the tool:

<img width="2904" height="1220" alt="Image" src="https://github.com/user-attachments/assets/e0503a51-591b-4b3c-b008-e203e8b721b6" />

After executing:

<img width="2912" height="1014" alt="Image" src="https://github.com/user-attachments/assets/4965362c-a294-4b8e-aef2-d677a2ebc3e4" />

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation

mgascam committed:
Merge branch 'develop' into cancel-auth-fee-remediation
